airflow/models/dag.py

```diff
@@ -368,8 +369,11 @@ def __init__(
                 DeprecationWarning,
                 stacklevel=2,
             )
-
-        validate_key(dag_id)
+        
+        if not is_ascii(dag_id):
+            # slugify dag id
+            dag_id = slugify(dag_id, lowercase=False)
```

Comment:
Is it true that the plan here is to allow your dag code to have one `dag_id` defined with non-ascii characters, but in the metastore, and therefore in the UI, you would have a _different_ dag id, with the value ascii-ified?  this seems undesirable to me.  seems we should either remove the ascii requirement or do nothing, so that code always agrees with metastore.

Reply:
@dstandish - missed the comment - but yeah that's the plan - to have different ids in the db. 

The problem is that current we have that limitation (all dag IDs are ASCII). And that in some cases (MySQL) we will not be able to get rid of that (UTF-8mb4 indexes are too big even for MySQL 8).

So yeah i t means that in UI the I'd will be different. I was pretty ok so far with this but your point might be valid - i am not sure really how bad it would be for those IDs to be different. I can't immediately see any obvious problem here (of course when you integrate with it via API etc. You need to use different I'd but when you query/list etc. You will see the slugified dag id).

Why do you think it will be a problem ?

Reply:
> I can't immediately see any obvious problem here (of course when you integrate with it via API etc. You need to use different I'd but when you query/list etc. You will see the slugified dag id).
> 
> Why do you think it will be a problem ?

So the issue I see with it is consistency.  As a data engineer, a lot of the job is managing chaos -- you want things to be organized and predictable and consistent.  And so the obvious scenario that comes to mind is, let's say there's a dag failure you've observed in the UI.  The dag_id in the UI will be `dag_Ni-Hao-_czesc`.  It's my job to fix that dag.  So I do `cmd+f` and search for it.  But I won't find it.  Because in _the code_ it is defined as `dag_你好_cześć`.

So this would make a data engineer's job more challenging.

If full unicode support is not practical in all databases, perhaps it could be a setting `allow_unicode_dag_ids` or something, and leave it up to the user to determine whether their database is happy enough with arbitrary unicode dag ids?

Reply:
Yeah, I hear you @dstandish. And I am indeed more convinced we should probably try to do as you say (the problem with unicode support for MySQL is mostly about very obscure characters - one that I know for sure is that one that someone tried to use in the past and caused some mysql issues: 🦅. Quite unlikely to be useful for engineering consistency.

But this approach is a bit risly. We would  have to modify some basic premise we have with the IDs.
I wonder what's the reason we have this (@Melodie97 - this is the reason why '*' is not allowed). I have a feeling that we need to understand why we have those limitations on dag id in the first place. 

```
KEY_REGEX = re.compile(r'^[\w.-]+$')
GROUP_KEY_REGEX = re.compile(r'^[\w-]+$')


def validate_key(k: str, max_length: int = 250):
    """Validates value used as a key."""
    if not isinstance(k, str):
        raise TypeError(f"The key has to be a string and is {type(k)}:{k}")
    if len(k) > max_length:
        raise AirflowException(f"The key has to be less than {max_length} characters")
    if not KEY_REGEX.match(k):
        raise AirflowException(
            f"The key ({k}) has to be made of alphanumeric characters, dashes, "
            f"dots and underscores exclusively"
        )


def validate_group_key(k: str, max_length: int = 200):
    """Validates value used as a group key."""
    if not isinstance(k, str):
        raise TypeError(f"The key has to be a string and is {type(k)}:{k}")
    if len(k) > max_length:
        raise AirflowException(f"The key has to be less than {max_length} characters")
    if not GROUP_KEY_REGEX.match(k):
        raise AirflowException(
            f"The key ({k!r}) has to be made of alphanumeric characters, dashes and underscores exclusively"
        )
```

BTW. @Melodie97  - this is one of the examples where seemingly simple change sparked discussion and it's quite unlikely we come to a conclusion quickly :). So rather than fix it for Outreachy it's good if you rather contribute to the discussion (if you can provide your opinion on that and got enough understanding (or maybe take a look at the code to try to figure out why we have the limitation at all. That STILL is a contribution (even if it is unlikely to produce a merged change soon).

Reply:
https://docs.python.org/3/library/stdtypes.html?highlight=isidentifier#str.isidentifier is probably of use here.

Reply:
Yeah. Same problem as with python_slugify. But I think the more important question is - WHY we are limitng ourselves now. Is there any reason (besides uf8mb4 causing too large index) why we are limited by the regexp patterns we have?  I think I might know why we should not use "." - because of the subdags - but not about allowing only small set of ASCII characters?). 

Does anyone know a good reason ?

Reply:
I know of no good reason to forbid unicode chars (and have idly proposed expanding it to support unicode in one of the linked issues)

Reply:
One of the reasons mentioned in slack discussion (@eladkal) was (and this might be actually the root cause) is that statsd does not support non-ascii characters.

Reply:
Maybe the slugify belongs in Statsd then? As I think you(?) mentioned OpenTelemetry doesn't have this limitation?